client_connection/manager_test.go

```diff
@@ -0,0 +1,141 @@
+	"github.com/stretchr/testify/suite"
+	"sync"
+	"testing"
+)
+
+type test_context struct {
+	suite.Suite
+	connManager         *connectionManager
+	fakeDiscoveryClient server.Client
+	fakeOpenVpn         *fake_openvpn_client
+}
+
+func (tc *test_context) SetupTest() {
+
+	tc.fakeDiscoveryClient = server.NewClientFake()
+
+	serviceProposal := service_discovery.NewServiceProposal(identity.FromAddress("vpn-node-1"), dto.Contact{})
+	tc.fakeDiscoveryClient.NodeRegister(serviceProposal)
+
+	dialogEstablisherFactory := func(identity identity.Identity) communication.DialogEstablisher {
+		return &fake_dialog{}
+	}
+
+	tc.fakeOpenVpn = &fake_openvpn_client{make(chan int, 1), nil}
+	fakeVpnClientFactory := func(vpnSession *session.VpnSession, session *mysterium_api_client.Session) (openvpn.Client, error) {
+		return tc.fakeOpenVpn, nil
+	}
+
+	tc.connManager = NewManager(tc.fakeDiscoveryClient, dialogEstablisherFactory, fakeVpnClientFactory)
+}
+
+func (tc *test_context) TestWhenNoConnectionIsMadeStatusIsNotConnected() {
+	assert.Equal(tc.T(), ConnectionStatus{NOT_CONNECTED, "", nil}, tc.connManager.Status())
+}
+
+func (tc *test_context) TestOnConnectErrorStatusIsNotConnectedAndLastErrorIsSet() {
+	fatalVpnError := errors.New("fatal connection error")
+	tc.fakeOpenVpn.onConnectReturnError = fatalVpnError
+	tc.fakeOpenVpn.resumeStart()
+
+	assert.Error(tc.T(), tc.connManager.Connect(identity.FromAddress("identity-1"), "vpn-node-1"))
+	assert.Equal(tc.T(), ConnectionStatus{NOT_CONNECTED, "", fatalVpnError}, tc.connManager.Status())
+}
+
+func (tc *test_context) TestWhenManagerMadeConnectionStatusReturnsConnectedStateAndSessionId() {
+	tc.fakeOpenVpn.resumeStart()
+
+	err := tc.connManager.Connect(identity.FromAddress("identity-1"), "vpn-node-1")
+
+	assert.NoError(tc.T(), err)
+	assert.Equal(tc.T(), ConnectionStatus{CONNECTED, "vpn-node-1-session", nil}, tc.connManager.Status())
+}
+
+func (tc *test_context) TestStatusReportsConnectingWhenConnectionIsInProgress() {
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		wg.Done()
+		tc.connManager.Connect(identity.FromAddress("identity-1"), "vpn-node-1")
+	}()
+	//wait for go function actually start, to avoid race condition, when we query Status before Connect call even begins.
+	wg.Wait()
+	assert.Equal(tc.T(), ConnectionStatus{NEGOTIATING, "", nil}, tc.connManager.Status())
+	tc.fakeOpenVpn.resumeStart()
+}
+
+func (tc *test_context) TearDownTest() {
+
+}
+
+func TestConnectionManagerSuite(t *testing.T) {
+	suite.Run(t, new(test_context))
+}
+
+type fake_openvpn_client struct {
+	connectionDelay      chan int
+	onConnectReturnError error
+}
+
+func (foc *fake_openvpn_client) resumeStart() {
+	foc.connectionDelay <- 1
+}
+
+func (foc *fake_openvpn_client) Start() error {
+	<-foc.connectionDelay
+	return foc.onConnectReturnError
+}
+
+func (foc *fake_openvpn_client) Wait() error {
+	return nil
+}
+
+func (foc *fake_openvpn_client) Stop() error {
+	return nil
+}
+
+type fake_dialog struct {
```

Comment:
I would add this Fake next to `communication` package, so that others could reuse it.
Or we can do it on next case

Reply:
In my opinion, reusing these "fake" or "stub" implementations is not very practical, as these implementations are usually test case specific anyway and will differ from component to component testing needs.